linux/src/threadpool_linux.c

```diff
@@ -103,7 +130,40 @@ static void on_timer_callback(sigval_t timer_data)
     }
     else
     {
-        timer_instance->work_function(timer_instance->work_function_ctx);
+        if (interlocked_add(&timer_instance->timer_status, 0) == TIMER_ENABLED)
+        {
+            // Make sure this timer is 
+            if (interlocked_add(&timer_instance->timer_trigger, 0) == 0)
+            {
+                // Increment the timer trigger so we know it won't go again
+                interlocked_increment(&timer_instance->timer_trigger);
+                struct timespec ts;
+                if (clock_gettime(CLOCK_REALTIME, &ts) == -1)
+                {
+                    LogInfo("Failure getting time from clock");
+                }
+                else
+                {
+                    struct timespec result;
+                    timespec_diff_macro(ts, timer_instance->start_time, result);
+
+                    if (result.tv_sec - timer_instance->its.it_interval.tv_sec < 0)
+                    {
```

Comment:
So the test is: did line 141 (clock_gettime) happen before line 640 when we registered another timer.

Seems like this might work, but are we basing the resolution on seconds? Could the tv_sec difference be 0 but the tv_usec is negative? So we only have a resolution of seconds here?

Besides that, I'm wondering if there is still a race, since we are now checking 3 different things  to synchronize on (timer_status, timer_trigger, start_time).

Couldn't an "old" call to "on_timer_callback" from the system get suspended around line 141 (before calling gettime here) then resume "later" so that it appears valid (enough time has passed)? Possibly could be fixed by getting the time before checking our synchronization variables.

My suspicion is we should consider putting more info into "timer_status", maybe just have multiple states like UNUSED, STARTING, STARTED, FIRING? That might make it easier to synchronize here and in the timer start function.

I also think we need some kind of volatile access on start_time or we could read it in this function before/during setting it for a new timer. So we still read the old start time or we see an invalid time assuming it's large enough that the write isn't atomic.

I need to think about this a bit more (tomorrow)

Reply:
After some time I figured out the what happens with the unwanted callbacks.

Say you have a callback that fires at 01:00:23:05
And you call the timer_delete at 01:00:23:01

The race condition in the timer system is such that the delete doesn't get to cancel the timer triggering.  So you won't get a mystery callback after that initial callback firing.  This is why it only happens once in a very long while because the timer and the delete have to hit within a small window.  I don't know the size of the window for the delete call and the removing of the callback and the doc doesn't even attempt to give you a clue.  I figure that the difference in the timeout that you set and the callback should not be closer than a second (that's why I remove the microseconds from the equation).  Since this will only happen once I made it so you don't have to go through this code after the first iteration.  

I also made some changes due to helgrind giving me clues as to where I should add interlocks.

Reply:
I think my first actionable comment here is can we:

1. Combine `timer_status` and `timer_trigger` to one variable if possible
2. Get the current time earlier in `on_timer_callback`
3. Guard  writing to and reading `start_time` in some critical section (either just use the same interlocked with a "busy" state like a spin lock or something else)